pandas/tests/frame/test_constructors.py

```diff
@@ -2746,3 +2750,23 @@ def test_construction_from_set_raises(self):
         msg = "Set type is unordered"
         with pytest.raises(TypeError, match=msg):
             pd.DataFrame({"a": {1, 2, 3}})
+
+    @pytest.mark.parametrize("copy", [None, False, True])
+    def test_dict_nocopy(self, copy):
+        a = np.array([1, 2])
+        b = pd.array([1, 2])
+        df = pd.DataFrame({"a": a, "b": b}, copy=copy)
+        df.iloc[0, 0] = 0
+        df.iloc[0, 1] = 0
+
+        if copy is None:
+            # copy for ndarray, no copy for EA
+            assert a[0] == 1
+            assert b[0] == 0
+
+        elif copy:
+            assert a[0] == 1
+            assert b[0] == 1
+        else:
+            assert a[0] == 0
+            assert b[0] == 0
```

Comment:
@TomAugspurger ive just about gotten this working, but this last assertion is still failing.  The trouble is that the `df.iloc[0, 1] = 0` line ends up calling `ExtensionBlock.set_inplace`, which incorrectly sets an entire new array on the block (i.e. #35417)

Reply:
looks like this also breaks down if we add a second column with the same dtype as `a`, since it consolidates on the `iloc.__setitem__`